src/engine/Operation.cpp

```diff
@@ -100,7 +101,10 @@ shared_ptr<const ResultTable> Operation::getResult(bool isRoot,
     auto lock =
         getExecutionContext()->getQueryTreeCache().pinnedSizes().wlock();
     forAllDescendants([&lock](QueryExecutionTree* child) {
-      if (child->getType() == QueryExecutionTree::OperationType::SCAN &&
+      // TODO ideally Operation should not need to know about its subtypes
+      // so, it would be a good idea to move this into a virtual method
+      // that is no-op for all types except IndexScan.
+      if (std::dynamic_pointer_cast<IndexScan>(child->getRootOperation()) &&
           child->getResultWidth() == 1) {
```

Comment:
@joka921 

A very common pattern throughout the codebase is `std::dynamic_pointer_cast<IndexScan>(child->getRootOperation()) && child->getResultWidth() == 1` or `std::dynamic_pointer_cast<IndexScan>(child->getRootOperation()) && child->getResultWidth() == 3`.

I don't understand what exactly this check is supposed to resemble, but in my opinion it would make sense to introduce a new member function to the `Operation` class (name to be determined) that returns false by default, and `getResultWidth() == 3|1` for `IndexScan`. This way we could avoid the dynamic_cast entirely (getResultWidth calls a virtual function in any case) and reduce duplicated code + avoid the dependency of `Operation.cpp` on `IndexScan.h`

Reply:
1. All of this code is somewhat broken anyway as it has to use the `numVariables` function and not `getResultWidth()`, so it is good to refactor this.

2. I like the idea, what about `isIndexScanWithNumVariables(size_t numVariables)`, which is then typically called with `1`and `3`?

3. I think we could still implement this function directly non-virtual using a dynamic cast in Operation.cpp, because we have to know about the children anyway conceptually:). But I have no strong opinion, just choose something.

Reply:
I think this can still be part of this PR.